### DISCUSSION TITLE
migrate mono-repo packages to own repositories

### DISCUSSION THREAD

**Discussion by bdougie** [octoherd/octoherd#20] [Ideas]
I had a conversation with the [storybook](https://github.com/storybookjs/storybook) team this morning and they mentioned an issue with sunsetting packages out of their main monorepo. They still want the package to existing, but need to move the contents and issues into its own repo. 

Curious if a script could be made to transfer folders to their own repo. Issues labeled for that package could also be transferred in a similar way. 

cc/ @shilman @ndelangen in case more context is needed

**Comment by gr2m** [comment]
Happy to help building that script, sounds like use case that fits @octoherd's current scope. Is there any place I can find out more what you want to do? Could it all be done programmatically with existing data or would you need to prompt to enter e.g. the repository name for each package you want to migrate out of the monorepo?

**Comment by ndelangen** [comment]
This would make sense to me:

An extract command that would ask:
- the directory to extract
- where to extract it to on fs
- url for repository
- tags to move issues to new repository

**Comment by gr2m** [comment]
I assume the script would do the following 

1. iterate through all folders in https://github.com/storybookjs/storybook/tree/next/addons
2. Create a new repository, e.g. for [`/addons/a11y`](https://github.com/storybookjs/storybook/tree/next/addons/a11y) it would create https://github.com/storybookjs/a11y
3. Move all contents to the new repository
4. Update `package.json` to point to the new repository
5. Move all issues with the label `addon: a11y`

I assume you'd also want to create additional files for testing/building/CI?

Does that sound right? Do all the packages have the 1:1 mapping of folder name to label name?

**Comment by ndelangen** [comment]
> I assume you'd also want to create additional files for testing/building/CI?

Yeah this is the hard part, but also the part where we can shed some of the monorepo weight.. We should not impose the whole monorepo building & testing setup, on these hopefully leaner repositories.

> Does that sound right? Do all the packages have the 1:1 mapping of folder name to label name?

yeah pretty much!

**Comment by gr2m** [comment]
An Octokit script wouldn't interact with the local file system, so it wouldn't run any git commands. It would work directly against GitHub's APIs. 

I think with git there are ways to git history and preserve authorship of the a directory, when moved into a new repository. That would not be possible with Octoherd. Would that be a problem?

The core bit of this functionality would be a folder copy feature, from one repository to another. That would be useful outside of Octoherd, it can be implemented as an [Octokit plugin](https://github.com/octokit/core.js#plugins). The plugin can later be used in the Octoherd script. I'd start with that.

**Reply by ndelangen** [reply]
yeah loosing git history would be a problem, I'm afraid..

**Reply by gr2m** [reply]
In that case I'm afraid it won't be a good fit. A script that runs against the FS and uses `git` can do this, but that's out of scope for @octoherd
